src/libos/src/vm/page_tracker.rs

```diff
@@ -0,0 +1,504 @@
+                            tracker_start_addr + s * PAGE_SIZE,
+                            PAGE_SIZE * (e - s + 1),
+                        )
+                        .unwrap();
+                        ret.push(committed_range);
+                        // Reset markers
+                        start = None;
+                        end = None;
+                    }
+                    _ => {
+                        unreachable!()
+                    }
+                }
+                i += 1;
+            }
+        }
+
+        let total_size = ret.iter().fold(0, |a, b| a + b.size());
+        if committed {
+            trace!("get committed ranges = {:?}", ret);
+            debug_assert!(total_size == self.inner.count_ones() * PAGE_SIZE);
+        } else {
+            trace!("get uncommitted ranges = {:?}", ret);
+            debug_assert!(total_size == self.inner.count_zeros() * PAGE_SIZE);
+        }
+
+        ret
+    }
+
+    pub fn split_from_new_start(&mut self, new_start: usize) {
+        debug_assert!(self.range.start() <= new_start && new_start < self.range.end());
+
+        let split_idx = (new_start - self.range.start()) / PAGE_SIZE;
+        let new_inner = self.inner.split_off(split_idx);
+
+        trace!(
+            "old range= {:?}, new_start = {:x}, idx = {:?}",
+            self.range,
+            new_start,
+            split_idx
+        );
+
+        self.inner = new_inner;
+        if self.inner.all() {
+            self.fully_committed = true;
+        }
+
+        self.range.set_start(new_start);
+    }
+
+    // Commit memory for the whole current VMA (VMATracker)
+    pub fn commit_current_vma(&mut self, need_update_global: bool) -> Result<()> {
+        debug_assert!(self.type_ == TrackerType::VMATracker);
+
+        if self.is_fully_committed() {
+            return Ok(());
+        }
+
+        // Commit EPC
+        if self.is_reserved_only() {
+            vm_epc::commit_epc_for_user_space(self.range().start(), self.range().size()).unwrap();
+        } else {
+            debug_assert!(self.is_partially_committed());
+            let uncommitted_ranges = self.get_ranges(false);
+            for range in uncommitted_ranges {
+                vm_epc::commit_epc_for_user_space(range.start(), range.size()).unwrap();
+            }
+        }
+
+        // Update the tracker
+        self.inner.fill(true);
+        self.fully_committed = true;
+
+        if need_update_global {
+            self.update_pages_for_global_tracker(self.range().start(), self.range().size());
+        }
+
+        Ok(())
+    }
+
+    // Commit memory of a specific range for the current VMA (VMATracker). The range should be verified by caller.
+    pub fn commit_range_for_current_vma(&mut self, range: &VMRange) -> Result<()> {
+        debug_assert!(self.type_ == TrackerType::VMATracker);
+
+        vm_epc::commit_epc_for_user_space(range.start(), range.size())?;
```

Comment:
In this function, the intersection range actually equals the parameter `range`, which is guaranteed by the caller. Here it is different from `commit_pages_internal` below. Because `commit_pages_internal` is used by both `GlobalTracker` and `VMATracker`.  Thus, the range specified for `commit_pages_internal` can be larger or intersect with self's range.

Reply:
`commit_range_for_current_vma` is a public method of `PageTracker`, so I think this method should guarantee that only the range of intersection is processed, not the caller.

Reply:
Add debug_assert.